---
title: "Seurat tutorial"
author: "Natalie Davidson"
date: "6/23/2025"
output: github_document
---

This tutorial is based off of the following references:
- https://biostatsquid.com/seurat-objects-explained/
- https://biostatsquid.com/single-cell-easy-tutorial-with-seurat-v5/
- https://satijalab.org/seurat/articles/pbmc3k_tutorial


```{r setup, fig.align='center', echo=TRUE, warning=FALSE, message=F}

# Loading relevant libraries 
library(Seurat)

#install.packages("devtools")
#devtools::install_github('satijalab/seurat-data')
library(SeuratData) # we'll be using a dataset from seurat data https://github.com/satijalab/seurat-data 

library(tidyverse)
library(pheatmap)
library(patchwork)

# Install dataset. only need to run this once..
#InstallData("pbmc3k")

pbmc <- LoadData("pbmc3k")


```


```{r look_at_assays, fig.align='center', echo=TRUE, warning=FALSE, message=F}

# Counts contains the raw counts
pbmc@assays$RNA$counts[c("CD3D", "TCL1A", "MS4A1"), 1:5]

# data should contain the normalised counts (but we haven't normalised yet!)
pbmc@assays$RNA$data[c("CD3D", "TCL1A", "MS4A1"), 1:5]


```


```{r look_at_metadata, fig.align='center', echo=TRUE, warning=FALSE, message=F}

# The [[ operator can add columns to object metadata. This is a great place to stash QC stats

# nCount_RNA which is the number of reads in that cell,
# nFeature_RNA, which is the number of unique genes that were detected in that cell.

pbmc@meta.data %>% head()
pbmc$percent.mt <- PercentageFeatureSet(pbmc, pattern = "^MT-")
pbmc$percent.ribo <- PercentageFeatureSet(pbmc, pattern = "^RB-")
pbmc@meta.data %>% head()


```


```{r normalize, fig.align='center', echo=TRUE, warning=FALSE, message=F}

pbmc <- NormalizeData(object = pbmc)


```


```{r get_variable_features, fig.align='center', echo=TRUE, warning=FALSE, message=F}

pbmc <- FindVariableFeatures(object = pbmc)
pbmc@assays$RNA@meta.data[21:31,]

```

```{r check_variable_features, fig.align='center', echo=TRUE, warning=FALSE, message=F}

table(pbmc@assays$RNA@meta.data$vf_vst_counts_variable) # How many variable genes? This can be changed in FindVariableFeatures()
pbmc@assays$RNA@meta.data$var.features %>% tail(50)

```

```{r plot_variable_features, fig.align='center', echo=TRUE, warning=FALSE, message=F}

VariableFeaturePlot(pbmc) 

```


```{r scale_variable_features, fig.align='center', echo=TRUE, warning=FALSE, message=F}

pbmc <- ScaleData(object = pbmc)
```


```{r compare_norm, fig.align='center', echo=TRUE, warning=FALSE, message=F}

pbmc@assays$RNA$counts[c("CD8A", "TCL1A", "MS4A1"), 20:25]
pbmc@assays$RNA$data[c("CD8A", "TCL1A", "MS4A1"), 20:25]
pbmc@assays$RNA$scale.data[c("TCL1A", "MS4A1"), 20:25]

```
